arangod/Aql/OptimizerRulePropagateConstantAttributes.cpp

```diff
@@ -0,0 +1,302 @@
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Max Neunhoeffer
+/// @author Jan Steemann
+////////////////////////////////////////////////////////////////////////////////
+
+#include "OptimizerRules.h"
+#include "Aql/Expression.h"
+#include "Aql/Collection.h"
+#include "Aql/Optimizer.h"
+#include "VocBase/LogicalCollection.h"
+#include "Cluster/ServerState.h"
+#include "Logger/LogMacros.h"
+#include "OptimizerUtils.h"
+
+using namespace arangodb;
+using namespace arangodb::aql;
+using namespace arangodb::containers;
+using EN = arangodb::aql::ExecutionNode;
+
+namespace {
+class PropagateConstantAttributesHelper {
+ public:
+  explicit PropagateConstantAttributesHelper(ExecutionPlan* plan)
+      : _plan(plan), _modified(false) {}
+
+  bool modified() const { return _modified; }
+
+  /// @brief inspects a plan and propagates constant values in expressions
+  void propagateConstants() {
+    LOG_DEVEL << "PROPAGATE CONSTANTS";
+    containers::SmallVector<ExecutionNode*, 8> nodes;
+    _plan->findNodesOfType(nodes, EN::FILTER, true);
+
+    for (auto* node = _plan->root()->getSingleton(); node != nullptr;
+         node = node->getFirstParent()) {
+      if (node->getType() != EN::FILTER) {
+        continue;
+      }
+
+      auto fn = ExecutionNode::castTo<FilterNode const*>(node);
+      auto setter = _plan->getVarSetBy(fn->inVariable()->id);
+      if (setter != nullptr && setter->getType() == EN::CALCULATION) {
+        auto cn = ExecutionNode::castTo<CalculationNode*>(setter);
+        auto expression = cn->expression();
+
+        insertConstantAttributes(const_cast<AstNode*>(expression->node()));
+        collectConstantAttributes(const_cast<AstNode*>(expression->node()));
+      }
+    }
+  }
+
+ private:
+  AstNode const* getConstant(Variable const* variable,
+                             std::string const& attribute) const {
+    auto it = _constants.find(variable);
+
+    if (it == _constants.end()) {
+      return nullptr;
+    }
+
+    auto it2 = (*it).second.find(attribute);
+
+    if (it2 == (*it).second.end()) {
+      return nullptr;
+    }
+
+    return (*it2).second;
+  }
+
+  /// @brief inspects an expression (recursively) and notes constant attribute
+  /// values so they can be propagated later
+  void collectConstantAttributes(AstNode* node) {
+    if (node == nullptr) {
+      return;
+    }
+
+    LOG_DEVEL << node->toString();
+
```

Comment:
I guess all these `LOG_DEVEL` should be removed, or changed to some `LOG_DEVEL_RULE` (as in other optimizer rules)